sketch/tensor_slide.hpp

```diff
@@ -53,44 +55,44 @@ class TensorSlide : public Tensor<seq_type> {
         auto T1 = new3D<double>(tup_len + 2, tup_len + 1, this->sketch_size, 0);
         auto T2 = new3D<double>(tup_len + 2, tup_len + 1, this->sketch_size, 0);
 
-        for (size_t p = 0; p <= tup_len; p++) {
+        for (uint32_t p = 0; p <= tup_len; p++) {
             T1[p + 1][p][0] = 1;
         }
 
         // T[p][q] at step i represents the sketch for seq[i-w+1]...seq[i] when only using hash
         // functions 1<=p,p+1,...q<=t, where t is the sketch size
-        for (size_t i = 0; i < seq.size(); i++) {
-            for (size_t p = 1; p <= tup_len; p++) {
+        for (uint32_t i = 0; i < seq.size(); i++) {
+            for (uint32_t p = 1; p <= tup_len; p++) {
                 // q-p must be smaller than i, hence the min in the condition
-                for (uint32_t q = std::min(p + i, (size_t)tup_len); q >= p; q--) {
+                for (uint32_t q = std::min(p + i, (uint32_t)tup_len); q >= p; q--) {
                     double z = (double)(q - p + 1) / std::min(i + 1, win_len + 1);
                     auto r = hashes[q - 1][seq[i]];
                     bool s = signs[q - 1][seq[i]];
                     if (s) {
-                        T1[p][q] = this->shift_sum(T1[p][q], T1[p][q - 1], r, z);
-                        T2[p][q] = this->shift_sum(T2[p][q], T2[p][q - 1], r, z);
+                        this->shift_sum_inplace(T1[p][q], T1[p][q - 1], r, z);
+                        this->shift_sum_inplace(T2[p][q], T2[p][q - 1], r, z);
                     } else {
-                        T1[p][q] = this->shift_sum(T1[p][q], T2[p][q - 1], r, z);
-                        T2[p][q] = this->shift_sum(T2[p][q], T1[p][q - 1], r, z);
+                        this->shift_sum_inplace(T1[p][q], T2[p][q - 1], r, z);
+                        this->shift_sum_inplace(T2[p][q], T1[p][q - 1], r, z);
                     }
                 }
             }
 
             if (i >= win_len) { // only start deleting from front after reaching #win_len
-                uint32_t ws = i - win_len; // the element to be removed from the sketch
-                for (size_t diff = 0; diff < tup_len; ++diff) {
-                    for (size_t p = 1; p <= tup_len - diff; p++) {
+                uint64_t ws = i - win_len; // the element to be removed from the sketch
```

Comment:
leave this as a 32 bit integer. Give the compiler a chance to optimize this using SSE.